Adds ability to download remote files with file provisioner

artburkart opened the pull request:
I'm struggling with this one. The point was to add the functionality that enables users to add urls to their sources list when using the "direction" => "upload" functionality, except I can't seem to get the downloads to work correctly yet. The file is created, but nothing is written to it. 

Once the bug is worked out, this PR would close #351 

I've commented what I think the problem areas are. Am I doing something obviously wrong? I'm brand new to golang, so I'm hoping this is the case. Cheers!

vtolstov commented on website/source/docs/provisioners/file.html.md:
```diff
@@ -30,17 +30,16 @@ The file provisioner can upload both single files and complete directories.
 
--   `sources` (array of strings) - A list of paths or directories to upload to
-    the machine. This is similar to the `source` option.
-
```
Why you remove sources? Why not convert source to array of string to sources and iterate over this array?
In this case packer template not grown with many files...

artburkart replied:
Hi @vtolstov, 

Good question. If I'm not mistaken, it looks like the `sources` config option is [something you implemented](https://github.com/mitchellh/packer/blob/feee19e4ed51e9045e4de6639b4b954f5ac37e58/provisioner/file/provisioner.go#L96) in Feb? It appears the docs for the feature may have been missed, which is why I added them here. I then removed them because I don't know what the intended behavior of `sources` is.

I'll explain why I feel this way. If I specify the following:

```json
{
    "sources": ["file1","path/file2"],
    "destination": "/tmp/"
}
```

I'd expect the files to get copied to the remote server like this:

1. /tmp/file1
2. /tmp/file2

What I get instead is this:

1. /tmp/tmp

Where /tmp/tmp contains the content of whichever file was copied last (/path/file2 in this case).

To be clear, I didn't remove the functionality of `sources`, I just continued to hide it as it always had been.

Good thing you raised it though because I was going to ask if it made sense to remove it or to improve it. If nothing else, I shouldn't be the one who documents it, since I don't know what its intended functionality was.

artburkart commented:
Hi, good question. If I'm not mistaken, it looks like the `sources` config option is [something you implemented](https://github.com/mitchellh/packer/blob/feee19e4ed51e9045e4de6639b4b954f5ac37e58/provisioner/file/provisioner.go#L96)? It doesn't appear to have ever been documented, which is why I added the docs for it here. 

I then removed the docs I added because the feature has undefined behavior and I thought it was being hidden for this reason.

I'll explain why the feature doesn't work as I'd expect. If I specify the following:

```json
"sources":["file1","path/file2"],
"destination":"/tmp/"
```

I'd expect the files to get copied to the remote server like this:

1. /tmp/file1
2. /tmp/file2

What I get instead is:

1. /tmp

Where /tmp contains the content of whichever file was copied last (/path/file2).

To be clear, I didn't remove the functionality, I just continued to hide it as it always had been.

Good thing you raised it because I was going to ask if it made sense to remove it or to improve it, since it's not documented and its behavior seems undefined.

artburkart commented:
This PR is still not ready. I wanted to add the ability to checksum files that are downloaded from a remote url, but didn't realize until today that that would mean I need to provide a `checksums` and `checksum_types` option. Now that I've spent a few hours trying to implement this, I think the file provisioner needs more love before this feature is implemented. 

For example, the current behavior when the following config is defined is slightly unexpected:

```json
{
    "sources": ["file1", "file2"],
    "destination": "/tmp/"
}
```

It results in a single file called /tmp/tmp with the content of file2.

If `checksums` was implemented, a whole slew of valid permutations ensue:

```json
{
    "source": "file1"
    "checksum": "sum1",
    "checksum_type": "md5"
}
```

```json
{
    "source": "file1"
    "checksums": ["sum1"]
    "checksum_type": "md5"
}
```

```json
{
    "sources": ["file1", "file2"]
    "checksum": "sum2"
    "checksums": ["sum1"]
    "checksum_types": ["md5", "sha256"]
}
```

```json
{
    "sources": ["file1", "file2"]
    "checksums": ["sum1", "sum2"],
    "checksum_type": "sha256",
    "checksum_types": ["md5"]
}
```

```json
{
    "source": "file1"
    "sources": ["file2"]
    "checksums": ["sum2", "sum1"]
    "checksum_types": ["sha256", "md5"]
}
```

I haven't covered all the permutations of 2 files, but I'll stop here for brevity. In the case of the last config, it would need to be documented that `source` gets appended to the end of `sources` (which is why sum2 comes before sum1 in `checksums`). 

Unless there's a situation where it's useful to scp multiple files to the same file location on a remote server, I'd argue the `sources` option is confusing and maybe not useful enough for the permutations it causes. Since `sources` has never been documented on the [file uploads page](https://www.packer.io/docs/provisioners/file.html) online, maybe it makes sense to remove it in favor of a code state with fewer permutations?

artburkart commented:
Thinking about this some more, maybe what this provisioner needs is something similar to the `ami_block_device_mappings` config. If `sources` was a list of configs rather than a list of strings, there is a lot more simplicity around how things can be processed, yet the flexibility and DRYness is maintained.

Something like:

```json
{
	"sources": [
		{
			"source": "/path/to/file1",
			"direction": "upload",
			"destination": "/somewhere/only/we/know"
		},
		{
			"source": "https://somewhere.example.com/file2",
			"checksum": "some_md5_checksum",
			"checksum_type": "md5",
			"direction": "upload",
			"destination": "/way/up/there"
		},
		{
			"source": "/tmp/file3",
			"direction": "download",
			"destination": "packer_cache/cuz_why_not"
		}
	]
}
```

Any thoughts or feedback is much appreciated :stuck_out_tongue:

vtolstov commented:
I'm prefer last solution with expanded sources

26 Окт 2016 г. 6:04 пользователь "Arthur Burkart" <notifications@github.com>
написал:

> Thinking about this some more, maybe what this provisioner needs is
> something similar to the ami_block_device_mappings config. If sources was
> a list of configs rather than a list of strings, there is a lot more
> simplicity around how things can be processed, yet the flexibility and
> DRYness is maintained.
>
> Something like:
>
> {
>     "sources": [
>         {
>             "source": "/path/to/file1",
>             "direction": "upload",
>             "destination": "/somewhere/only/we/know"
>         },
>         {
>             "source": "https://somewhere.example.com/file2",
>             "checksum": "some_md5_checksum",
>             "checksum_type": "md5",
>             "direction": "upload",
>             "destination": "/way/up/there"
>         },
>         {
>             "source": "/tmp/file3",
>             "direction": "download",
>             "destination": "packer_cache/cuz_why_not"
>         }
>     ]
> }
>
> Any thoughts or feedback is much appreciated 😛
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mitchellh/packer/pull/4049#issuecomment-256237752>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAdYG2Onh9bs5bmZDYet-JQdNDYMSbRuks5q3sNEgaJpZM4KeOpO>
> .
>

artburkart commented:
> I'm prefer last solution with expanded sources

If other people indicate an interest in this implementation, I'll consider executing it.

lfarnell commented:
@artburkart I personally think the solution with expanded sources should be fine.

artburkart commented:
I don't think my proposed solution is convenient because it would require specifying `destination` over and over again. I don't think I'll implement it.
